internal/manage/devices/download.go

```diff
@@ -0,0 +1,337 @@
+package devices
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/csv"
+	"fmt"
+	"iter"
+	"maps"
+	"net/http"
+	"net/url"
+	"slices"
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protopath"
+	"google.golang.org/protobuf/reflect/protorange"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	"github.com/smart-core-os/sc-api/go/traits"
+	"github.com/smart-core-os/sc-golang/pkg/resource"
+	"github.com/smart-core-os/sc-golang/pkg/trait"
+	"github.com/vanti-dev/sc-bos/pkg/gen"
+)
+
+//go:generate protomod protoc -- -I . -I ../../../proto --go_out=paths=source_relative:. download.proto
+
+func (s *Server) GetDownloadDevicesUrl(_ context.Context, request *gen.GetDownloadDevicesUrlRequest) (*gen.DownloadDevicesUrl, error) {
+	// validate
+	switch request.MediaType {
+	case "":
+		request.MediaType = "text/csv"
+	case "text/csv": // supported
+	default:
+		return nil, status.Errorf(codes.InvalidArgument, "unsupported media type %q", request.MediaType)
+	}
+
+	tokenData := &DownloadToken{
+		Request: request,
+	}
+	tokenStr, err := encodeDownloadToken(tokenData)
+	if err != nil {
+		return nil, err
+	}
+	u := s.downloadUrlBase
+	if err := s.encodeDownloadUrlToken(&u, tokenStr); err != nil {
+		return nil, err
+	}
+	return &gen.DownloadDevicesUrl{
+		Url: u.String(),
+	}, nil
+}
+
+// DownloadDevicesHTTPHandler responds to HTTP request urls returned by GetDownloadDevicesUrl.
+func (s *Server) DownloadDevicesHTTPHandler(w http.ResponseWriter, r *http.Request) {
+	// 1. parse and validate the request
+	tokenStr, err := s.decodeDownloadUrlToken(r)
+	if err != nil {
+		http.Error(w, "invalid download token", http.StatusBadRequest)
+		return
+	}
+	token, err := decodeDownloadToken(tokenStr)
+	if err != nil {
+		http.Error(w, "corrupted download token", http.StatusBadRequest)
+		return
+	}
+
+	// 2. collect header related information
+	deviceList := s.node.ListAllMetadata(
+		resource.WithInclude(func(id string, item proto.Message) bool {
+			if item == nil {
+				return false
+			}
+			device := &gen.Device{
+				Name:     id,
+				Metadata: item.(*traits.Metadata),
+			}
+			return deviceMatchesQuery(token.Request.Query, device)
+		}),
+	)
+	usefulDevices := func() iter.Seq[*traits.Metadata] {
+		return func(yield func(*traits.Metadata) bool) {
+			for _, d := range deviceList {
+				if len(d.Traits) == 0 {
+					continue
+				}
+				if proto.Equal(d, &traits.Metadata{Name: d.Name, Traits: []*traits.TraitMetadata{{Name: string(trait.Metadata)}}}) {
+					continue // skip boring devices
+				}
```

Comment:
What's the rationale for hiding these devices?

Reply:
Mostly it's to remove rows from the CSV that have no data other than the name of the device - typically these are `automations` and things like that. It might be that future updates add options to include them, but for the cases here they just filled the CSV with (almost) empty rows